ocular/src/chain/client/airdrop.rs

```diff
@@ -0,0 +1,182 @@
+};
+use bip32::Mnemonic;
+use cosmos_sdk_proto::cosmos::authz::v1beta1::Grant;
+use cosmrs::{
+    bank::{MsgMultiSend, MsgSend},
+    rpc,
+    tx::Msg,
+    AccountId,
+};
+
+use prost::Message;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, time::SystemTime};
+use std::{fs, path::Path, str::FromStr};
+use tendermint_rpc::endpoint::broadcast::tx_commit::Response;
+use uuid::Uuid;
+
+use super::ChainClient;
+
+const MSG_MULTI_SEND_URL: &str = "/cosmos.bank.v1beta1.MsgMultiSend";
+const GENERIC_AUTHORIZATION_URL: &str = "/cosmos.authz.v1beta1.GenericAuthorization";
+
+impl ChainClient {
+    pub async fn verify_multi_send_grant(
+        &mut self,
+        granter: AccountId,
+        grantee: AccountId,
+    ) -> Result<bool, ChainClientError> {
+        // Verify grant exists for grantee from granter for MsgSend
+        let res = self
+            .query_authz_grant(granter.as_ref(), grantee.as_ref(), MSG_MULTI_SEND_URL)
+            .await?;
+
+        // If any grants meet the following criteria we can be confident the transaction is authorized:
+        // 1. The grant either has no expiration, or an expiration with more than 60 seconds remaining.
+        // 2. The grant contains a generic authorization
+        Ok(res.grants.iter().any(|g| {
+            if g.expiration.is_some() {
+                let expiration = g.expiration.clone().unwrap();
+                let cutoff = i64::try_from(
+                    SystemTime::now()
+                        .duration_since(SystemTime::UNIX_EPOCH)
+                        .unwrap()
+                        .as_secs()
+                        + 60,
+                )
+                .expect("failed to derive time from system clock");
+
+                if expiration.seconds <= cutoff {
+                    return false;
+                }
+            }
+            if g.authorization.is_none() {
+                return false;
+            }
+            // I don't actually thing this is a necessary check as there is no way to specify
+            // authorization for MultiSend without using a generic one
+            let authorization = g.authorization.clone().unwrap();
+            if authorization.type_url.as_str() != GENERIC_AUTHORIZATION_URL {
+                return false;
+            }
+
+            true
+        }))
+    }
+
+    pub async fn execute_delegated_airdrop(
+        &mut self,
+        granter: AccountInfo,
+        grantee: AccountInfo,
+        fee_payer: Option<AccountId>,
+        fee_granter: Option<AccountId>,
+        payments: Vec<Payment>,
+        tx_metadata: Option<TxMetadata>,
+    ) -> Result<Response, ChainClientError> {
+        self.verify_multi_send_grant(granter.id.clone(), grantee.id.clone())
+            .await?;
+        let (inputs, outputs) =
+            multi_send_args_from_payments(granter.id.as_ref().to_string(), payments);
+        let msgs: Vec<prost_types::Any> = vec![MsgMultiSend {
+            inputs: inputs
```

Comment:
Interesting, do these not require individual tx metadata? How do memos and the like work here? Wondering what happens if a specific kind of memo for a message is required (e.g. https://thefipharmacist.com/memo-needed-for-crypto/, have another comment on this below)

Reply:
Right, I'm just wondering what the tx looks like for individual accts receiving funds, I'd assume they all receive no memo or the same memo as the base multisend, either of which would be issues for CEX addresses -- basically just a continuation of the other convo regarding power users vs non power/unaware receivers

Reply:
CEX addresses aren't eligible for airdrops in general. The way airdrops usually work in every situation I've seen is a snapshot is taken of wallets that participated somehow in some selected DeFi activities. Addresses in a CEX can never do this because you have to own the private key to connect your wallet to DeFi, but CEX wallets are custodial. Airdropping to wallets reported by users would be easy to scam. For these reasons I don't think there's a scenario where airdrops can be sent to a CEX address?

Reply:
Totally agree in the general case, however for our monthly airdrops we'll be letting airdropee's self report https://www.notion.so/Token-Distribution-82ec48d2a4da42f39f4c50ac54e27ccc

Reply:
Perhaps we can just verify self ownership in the google form they have to submit? We just have to be diligent about that when the time comes